man/man1/notmuch-search.1
.TH NOTMUCH-SEARCH 1 2013-12-30 "Notmuch 0.17"
.SH NAME
notmuch-search \- search for messages matching the given search terms
.SH SYNOPSIS

.B notmuch search
.RI  [  options "...] <" search-term ">..."

.SH DESCRIPTION

Search for messages matching the given search terms, and display as
results the threads containing the matched messages.

The output consists of one line per thread, giving a thread ID, the
date of the newest (or oldest, depending on the sort option) matched
message in the thread, the number of matched messages and total
messages in the thread, the names of all participants in the thread,
and the subject of the newest (or oldest) message.

See \fBnotmuch-search-terms\fR(7)
for details of the supported syntax for <search-terms>.

Supported options for
.B search
include
.RS 4
.TP 4
.BR \-\-format= ( json | sexp | text | text0 )

Presents the results in either JSON, S-Expressions, newline character
separated plain-text (default), or null character separated plain-text
(compatible with \fBxargs\fR(1) -0 option where available).
.RE

.RS 4
.TP 4
.BR \-\-format-version=N

Use the specified structured output format version.  This is intended
for programs that invoke \fBnotmuch\fR(1) internally.  If omitted, the
latest supported version will be used.
.RE

.RS 4
.TP 4
.B \-\-output=(summary|threads|messages|files|tags)

.RS 4
.TP 4
.B summary

Output a summary of each thread with any message matching the search
terms. The summary includes the thread ID, date, the number of
messages in the thread (both the number matched and the total number),
the authors of the thread and the subject.
.RE
.RS 4
.TP 4
.B threads

Output the thread IDs of all threads with any message matching the
search terms, either one per line (\-\-format=text), separated by null
characters (\-\-format=text0), as a JSON array (\-\-format=json), or
an S-Expression list (\-\-format=sexp).
.RE
.RS 4
.TP 4
.B messages

Output the message IDs of all messages matching the search terms,
either one per line (\-\-format=text), separated by null characters
(\-\-format=text0), as a JSON array (\-\-format=json), or as an
S-Expression list (\-\-format=sexp).
.RE
.RS 4
.TP 4
.B files

Output the filenames of all messages matching the search terms, either
one per line (\-\-format=text), separated by null characters
(\-\-format=text0), as a JSON array (\-\-format=json), or as an
S-Expression list (\-\-format=sexp).

Note that each message may have multiple filenames associated with it.
All of them are included in the output, unless limited with the
\-\-duplicate=N option.
.RE
.RS 4
.TP 4
.B tags

Output all tags that appear on any message matching the search terms,
either one per line (\-\-format=text), separated by null characters
(\-\-format=text0), as a JSON array (\-\-format=json), or as an
S-Expression list (\-\-format=sexp).
.RE
.RE

.RS 4
.TP 4
.BR \-\-sort= ( newest\-first | oldest\-first )

This option can be used to present results in either chronological order
.RB ( oldest\-first )
or reverse chronological order
.RB ( newest\-first ).

Note: The thread order will be distinct between these two options
(beyond being simply reversed). When sorting by
.B oldest\-first
the threads will be sorted by the oldest message in each thread, but
when sorting by
.B newest\-first
the threads will be sorted by the newest message in each thread.

By default, results will be displayed in reverse chronological order,
(that is, the newest results will be displayed first).
.RE

.RS 4
.TP 4
.BR \-\-offset=[\-]N

Skip displaying the first N results. With the leading '\-', start at the Nth
result from the end.
.RE

.RS 4
.TP 4
.BR \-\-limit=N

Limit the number of displayed results to N.
.RE

.RS 4
.TP 4
.BR \-\-exclude=(true|false|all|flag)

A message is called "excluded" if it matches at least one tag in
search.tag_exclude that does not appear explicitly in the search terms.
This option specifies whether to omit excluded messages in the search
process.

The default value,
.BR true ,
prevents excluded messages from matching the search terms.

.B all
additionally prevents excluded messages from appearing in displayed
results, in effect behaving as though the excluded messages do not exist.

.B false
allows excluded messages to match search terms and appear in displayed
results. Excluded messages are still marked in the relevant outputs.

.B flag
only has an effect when
.BR --output=summary .
The output is almost identical to
.BR false ,
but the "match count" is the number of matching non-excluded messages in the
thread, rather than the number of matching messages.
.RE

.RS 4
.TP 4
.BR \-\-duplicate=N

Effective with
.BR --output=files ,
output the Nth filename associated with each message matching the
query (N is 1-based). If N is greater than the number of files
associated with the message, don't print anything.

Note that this option is orthogonal with the
.BR folder:
search prefix. The prefix matches messages based on filenames. This
option filters filenames of the matching messages.
.RE

.SH EXIT STATUS

This command supports the following special exit status codes

.TP
.B 20
The requested format version is too old.
.TP
.B 21
The requested format version is too new.

.SH SEE ALSO

\fBnotmuch\fR(1), \fBnotmuch-config\fR(1), \fBnotmuch-count\fR(1),
\fBnotmuch-dump\fR(1), \fBnotmuch-hooks\fR(5),
\fBnotmuch-insert\fR(1), \fBnotmuch-new\fR(1),
\fBnotmuch-reply\fR(1), \fBnotmuch-restore\fR(1),
\fBnotmuch-search-terms\fR(7), \fBnotmuch-show\fR(1),
\fBnotmuch-tag\fR(1)


NEWS
Notmuch 0.17~rc4 (2013-12-28)
=============================

Incompatible change in SHA1 computation
---------------------------------------

Previously on big endian architectures like sparc and powerpc the
computation of SHA1 hashes was incorrect. This meant that messages
with overlong or missing message-ids were given different computed
message-ids than on more common little endian architectures like i386
and amd64.  If you use notmuch on a big endian architecture, you are
strongly advised to make a backup of your tags using `notmuch dump`
before this upgrade. You can locate the affected files using something
like:

    notmuch dump | \
      awk '/^notmuch-sha1-[0-9a-f]{40} / \
        {system("notmuch search --exclude=false --output=files id:" $1)}'

Command-Line Interface
----------------------

New options to better support handling duplicate messages

  If more than one message file is associated with a message-id,
  `notmuch search --output=files` will print all of them. A new
  `--duplicate=N` option can be used to specify which duplicate to
  print for each message.

  `notmuch count` now supports `--output=files` option to output the
  number of files associated with matching messages. This may be
  bigger than the number of matching messages due to duplicates
  (i.e. multiple files having the same message-id).

Improved `notmuch new` performance for unchanged folders

  `notmuch new` now skips over unchanged folders more efficiently,
  which can substantially improve the performance of checking for new
  mail in some situations (like NFS-mounted Maildirs).

`notmuch reply --format=text` RFC 2047-encodes headers

  Previously, this used a mix of standard MIME encoding for the reply
  body and UTF-8 for the headers.  Now, the text format reply template
  RFC 2047-encodes the headers, making the output a valid RFC 2822
  message.  The JSON/sexp format is unchanged.

`notmuch compact` command

  The new `compact` command exposes Xapian's compaction
  functionality through a more convenient interface than
  `xapian-compact`. `notmuch compact` will compact the database to a
  temporary location, optionally backup the original database, and
  move the compacted database into place.

Emacs Interface
---------------

`notmuch-tree` (formerly `notmuch-pick`) has been added to mainline

  `notmuch-tree` is a threaded message view for the emacs
  interface. Each message is one line in the results and the thread
  structure is shown using UTF-8 box drawing characters (similar to
  Mutt's threaded view). It comes between search and show in terms of
  amount of output and can be useful for viewing both single threads
  and multiple threads.

  Using `notmuch-tree`

  The main key entries to notmuch tree are

  'z' enter a query to view using notmuch tree (works in hello,
      search, show and tree mode itself)

  'Z' view the current query in tree notmuch tree (works from search
      and show)

  Once in tree mode, keybindings are mostly in line with the rest of
  notmuch and are all viewable with '?' as usual.

  Customising `notmuch-tree`

  `notmuch-tree` has several customisation variables. The most
  significant is the first notmuch-tree-show-out which determines the
  behaviour when selecting a message (with RET) in tree view. By
  default tree view uses a split window showing the single message in
  the bottom pane. However, if this option is set then it views the
  whole thread in the complete window jumping to the selected message
  in the thread. In either case command-prefix selects the other option.

Tagging threads in search is now race-free

  Previously, adding or removing a tag from a thread in a search
  buffer would affect messages that had arrived after the search was
  performed, resulting in, for example, archiving messages that were
  never seen.  Tagging now affects only the messages that were in the
  thread when the search was performed.

`notmuch-hello` refreshes when switching to the buffer

  The hello buffer now refreshes whenever you switch to the buffer,
  regardless of how you get there.  You can disable automatic
  refreshing by customizing `notmuch-hello-auto-refresh`.

Specific mini-buffer prompts for tagging operations

  When entering tags to add or remove, the mini-buffer prompt now
  indicates what operation will be performed (e.g., "Tag thread", "Tag
  message", etc).

Built-in help improvements

  Documentation for many commands has been improved, as displayed by
  `notmuch-help` (usually bound to "?").  The bindings listed by
  `notmuch-help` also now include descriptions of prefixed commands.

Quote replies as they are displayed in show view

  We now render the parts for reply quoting the same way they are
  rendered for show. At this time, the notable change is that replies
  to text/calendar are now pretty instead of raw vcalendar.

Fixed inconsistent use of configured search order

  All ways of interactively invoking search now honor the value of
  `notmuch-search-oldest-first`.

Common keymap for notmuch-wide bindings

  Several key bindings have been moved from mode-specific keymaps to
  the single `notmuch-common-keymap`, which is inherited by each
  notmuch mode.  If you've customized your key bindings, you may want
  to move some of them to the common keymap.

The `notmuch-tag` function now requires a list of tag changes

  For users who have scripted the Emacs interface: the `notmuch-tag`
  API has changed.  Previously, it accepted either a list of tag
  changes or a space-separated string of tag changes.  The latter is
  no longer supported and the function now returns nothing.

Fixed `notmuch-reply` putting reply in primary selection

  On emacs 24 notmuch-reply used to put the cited text into the
  primary selection (which could lead to inadvertently pasting this
  cited text elsewhere). Now the primary-selection is not changed.

Fixed `notmuch-show` invisible part handling


notmuch-search.c
/* notmuch - Not much of an email program, (just index and search)
 *
 * Copyright © 2009 Carl Worth
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/ .
 *
 * Author: Carl Worth <cworth@cworth.org>
 */

#include "notmuch-client.h"
#include "sprinter.h"

typedef enum {
    OUTPUT_SUMMARY,
    OUTPUT_THREADS,
    OUTPUT_MESSAGES,
    OUTPUT_FILES,
    OUTPUT_TAGS
} output_t;

static char *
sanitize_string (const void *ctx, const char *str)
{
    char *out, *loop;

    if (NULL == str)
	return NULL;

    loop = out = talloc_strdup (ctx, str);

    for (; *loop; loop++) {
	if ((unsigned char)(*loop) < 32)
	    *loop = '?';
    }
    return out;
}

static int
do_search_threads (sprinter_t *format,
		   notmuch_query_t *query,
		   notmuch_sort_t sort,
		   output_t output,
		   int offset,
		   int limit)
{
    notmuch_thread_t *thread;
    notmuch_threads_t *threads;
    notmuch_tags_t *tags;
    time_t date;
    int i;

    if (offset < 0) {
	offset += notmuch_query_count_threads (query);
	if (offset < 0)
	    offset = 0;
    }

    threads = notmuch_query_search_threads (query);
    if (threads == NULL)
	return 1;

    format->begin_list (format);

    for (i = 0;
	 notmuch_threads_valid (threads) && (limit < 0 || i < offset + limit);
	 notmuch_threads_move_to_next (threads), i++)
    {
	thread = notmuch_threads_get (threads);

	if (i < offset) {
	    notmuch_thread_destroy (thread);
	    continue;
	}

	if (output == OUTPUT_THREADS) {
	    format->set_prefix (format, "thread");
	    format->string (format,
			    notmuch_thread_get_thread_id (thread));
	    format->separator (format);
	} else { /* output == OUTPUT_SUMMARY */
	    void *ctx_quote = talloc_new (thread);
	    const char *authors = notmuch_thread_get_authors (thread);
	    const char *subject = notmuch_thread_get_subject (thread);
	    const char *thread_id = notmuch_thread_get_thread_id (thread);
	    int matched = notmuch_thread_get_matched_messages (thread);
	    int total = notmuch_thread_get_total_messages (thread);
	    const char *relative_date = NULL;
	    notmuch_bool_t first_tag = TRUE;

	    format->begin_map (format);

	    if (sort == NOTMUCH_SORT_OLDEST_FIRST)
		date = notmuch_thread_get_oldest_date (thread);
	    else
		date = notmuch_thread_get_newest_date (thread);

	    relative_date = notmuch_time_relative_date (ctx_quote, date);

	    if (format->is_text_printer) {
                /* Special case for the text formatter */
		printf ("thread:%s %12s [%d/%d] %s; %s (",
			thread_id,
			relative_date,
			matched,
			total,
			sanitize_string (ctx_quote, authors),
			sanitize_string (ctx_quote, subject));
	    } else { /* Structured Output */
		format->map_key (format, "thread");
		format->string (format, thread_id);
		format->map_key (format, "timestamp");
		format->integer (format, date);
		format->map_key (format, "date_relative");
		format->string (format, relative_date);
		format->map_key (format, "matched");
		format->integer (format, matched);
		format->map_key (format, "total");
		format->integer (format, total);
		format->map_key (format, "authors");
		format->string (format, authors);
		format->map_key (format, "subject");
		format->string (format, subject);
	    }

	    talloc_free (ctx_quote);

	    format->map_key (format, "tags");
	    format->begin_list (format);

	    for (tags = notmuch_thread_get_tags (thread);
		 notmuch_tags_valid (tags);
		 notmuch_tags_move_to_next (tags))
	    {
		const char *tag = notmuch_tags_get (tags);

		if (format->is_text_printer) {
                  /* Special case for the text formatter */
		    if (first_tag)
			first_tag = FALSE;
		    else
			fputc (' ', stdout);
		    fputs (tag, stdout);
		} else { /* Structured Output */
		    format->string (format, tag);
		}
	    }

	    if (format->is_text_printer)
		printf (")");

	    format->end (format);
	    format->end (format);
	    format->separator (format);
	}

	notmuch_thread_destroy (thread);
    }

    format->end (format);

    return 0;
}

static int
do_search_messages (sprinter_t *format,
		    notmuch_query_t *query,
		    output_t output,
		    int offset,
		    int limit,
		    int dupe)
{
    notmuch_message_t *message;
    notmuch_messages_t *messages;
    notmuch_filenames_t *filenames;
    int i;

    if (offset < 0) {
	offset += notmuch_query_count_messages (query);
	if (offset < 0)
	    offset = 0;
    }

    messages = notmuch_query_search_messages (query);
    if (messages == NULL)
	return 1;

    format->begin_list (format);

    for (i = 0;
	 notmuch_messages_valid (messages) && (limit < 0 || i < offset + limit);
	 notmuch_messages_move_to_next (messages), i++)
    {
	if (i < offset)
	    continue;

	message = notmuch_messages_get (messages);

	if (output == OUTPUT_FILES) {
	    int j;
	    filenames = notmuch_message_get_filenames (message);

	    for (j = 1;
		 notmuch_filenames_valid (filenames);
		 notmuch_filenames_move_to_next (filenames), j++)
	    {
		if (dupe < 0 || dupe == j) {
		    format->string (format, notmuch_filenames_get (filenames));
		    format->separator (format);
		}
	    }
	    
	    notmuch_filenames_destroy( filenames );

	} else { /* output == OUTPUT_MESSAGES */
	    format->set_prefix (format, "id");
	    format->string (format,
			    notmuch_message_get_message_id (message));
	    format->separator (format);
	}

	notmuch_message_destroy (message);
    }

    notmuch_messages_destroy (messages);

    format->end (format);

    return 0;
}

static int
do_search_tags (notmuch_database_t *notmuch,
		sprinter_t *format,
		notmuch_query_t *query)
{
    notmuch_messages_t *messages = NULL;
    notmuch_tags_t *tags;
    const char *tag;

    /* should the following only special case if no excluded terms
     * specified? */

    /* Special-case query of "*" for better performance. */
    if (strcmp (notmuch_query_get_query_string (query), "*") == 0) {
	tags = notmuch_database_get_all_tags (notmuch);
    } else {
	messages = notmuch_query_search_messages (query);
	if (messages == NULL)
	    return 1;

	tags = notmuch_messages_collect_tags (messages);
    }
    if (tags == NULL)
	return 1;

    format->begin_list (format);

    for (;
	 notmuch_tags_valid (tags);
	 notmuch_tags_move_to_next (tags))
    {
	tag = notmuch_tags_get (tags);

	format->string (format, tag);
	format->separator (format);

    }

    notmuch_tags_destroy (tags);

    if (messages)
	notmuch_messages_destroy (messages);

    format->end (format);

    return 0;
}

int
notmuch_search_command (notmuch_config_t *config, int argc, char *argv[])
{
    notmuch_database_t *notmuch;
    notmuch_query_t *query;
    char *query_str;
    notmuch_sort_t sort = NOTMUCH_SORT_NEWEST_FIRST;
    sprinter_t *format = NULL;
    int opt_index, ret;
    output_t output = OUTPUT_SUMMARY;
    int offset = 0;
    int limit = -1; /* unlimited */
    notmuch_exclude_t exclude = NOTMUCH_EXCLUDE_TRUE;
    int dupe = -1;
    unsigned int i;

    enum {
	NOTMUCH_FORMAT_JSON,
	NOTMUCH_FORMAT_TEXT,
	NOTMUCH_FORMAT_TEXT0,
	NOTMUCH_FORMAT_SEXP
    } format_sel = NOTMUCH_FORMAT_TEXT;

    notmuch_opt_desc_t options[] = {
	{ NOTMUCH_OPT_KEYWORD, &sort, "sort", 's',
	  (notmuch_keyword_t []){ { "oldest-first", NOTMUCH_SORT_OLDEST_FIRST },
				  { "newest-first", NOTMUCH_SORT_NEWEST_FIRST },
				  { 0, 0 } } },
	{ NOTMUCH_OPT_KEYWORD, &format_sel, "format", 'f',
	  (notmuch_keyword_t []){ { "json", NOTMUCH_FORMAT_JSON },
				  { "sexp", NOTMUCH_FORMAT_SEXP },
				  { "text", NOTMUCH_FORMAT_TEXT },
				  { "text0", NOTMUCH_FORMAT_TEXT0 },
				  { 0, 0 } } },
	{ NOTMUCH_OPT_INT, &notmuch_format_version, "format-version", 0, 0 },
	{ NOTMUCH_OPT_KEYWORD, &output, "output", 'o',
	  (notmuch_keyword_t []){ { "summary", OUTPUT_SUMMARY },
				  { "threads", OUTPUT_THREADS },
				  { "messages", OUTPUT_MESSAGES },
				  { "files", OUTPUT_FILES },
				  { "tags", OUTPUT_TAGS },
				  { 0, 0 } } },
        { NOTMUCH_OPT_KEYWORD, &exclude, "exclude", 'x',
          (notmuch_keyword_t []){ { "true", NOTMUCH_EXCLUDE_TRUE },
                                  { "false", NOTMUCH_EXCLUDE_FALSE },
                                  { "flag", NOTMUCH_EXCLUDE_FLAG },
                                  { "all", NOTMUCH_EXCLUDE_ALL },
                                  { 0, 0 } } },
	{ NOTMUCH_OPT_INT, &offset, "offset", 'O', 0 },
	{ NOTMUCH_OPT_INT, &limit, "limit", 'L', 0  },
	{ NOTMUCH_OPT_INT, &dupe, "duplicate", 'D', 0  },
	{ 0, 0, 0, 0, 0 }
    };

    opt_index = parse_arguments (argc, argv, options, 1);

    if (opt_index < 0) {
	return 1;
    }

    switch (format_sel) {
    case NOTMUCH_FORMAT_TEXT:
	format = sprinter_text_create (config, stdout);
	break;
    case NOTMUCH_FORMAT_TEXT0:
	if (output == OUTPUT_SUMMARY) {
	    fprintf (stderr, "Error: --format=text0 is not compatible with --output=summary.\n");
	    return 1;
	}
	format = sprinter_text0_create (config, stdout);
	break;
    case NOTMUCH_FORMAT_JSON:
	format = sprinter_json_create (config, stdout);
	break;
    case NOTMUCH_FORMAT_SEXP:
	format = sprinter_sexp_create (config, stdout);
	break;
    default:
	/* this should never happen */
	INTERNAL_ERROR("no output format selected");
    }

    notmuch_exit_if_unsupported_format ();

    if (notmuch_database_open (notmuch_config_get_database_path (config),
			       NOTMUCH_DATABASE_MODE_READ_ONLY, &notmuch))
	return 1;

    query_str = query_string_from_args (notmuch, argc-opt_index, argv+opt_index);
    if (query_str == NULL) {
	fprintf (stderr, "Out of memory.\n");
	return 1;
    }
    if (*query_str == '\0') {
	fprintf (stderr, "Error: notmuch search requires at least one search term.\n");
	return 1;
    }

    query = notmuch_query_create (notmuch, query_str);
    if (query == NULL) {
	fprintf (stderr, "Out of memory\n");
	return 1;
    }

    notmuch_query_set_sort (query, sort);

    if (exclude == NOTMUCH_EXCLUDE_FLAG && output != OUTPUT_SUMMARY) {
	/* If we are not doing summary output there is nowhere to
	 * print the excluded flag so fall back on including the
	 * excluded messages. */
	fprintf (stderr, "Warning: this output format cannot flag excluded messages.\n");
	exclude = NOTMUCH_EXCLUDE_FALSE;
    }

    if (exclude != NOTMUCH_EXCLUDE_FALSE) {
	const char **search_exclude_tags;
	size_t search_exclude_tags_length;

	search_exclude_tags = notmuch_config_get_search_exclude_tags
	    (config, &search_exclude_tags_length);
	for (i = 0; i < search_exclude_tags_length; i++)
	    notmuch_query_add_tag_exclude (query, search_exclude_tags[i]);
	notmuch_query_set_omit_excluded (query, exclude);
    }

    switch (output) {
    default:
    case OUTPUT_SUMMARY:
    case OUTPUT_THREADS:
	ret = do_search_threads (format, query, sort, output, offset, limit);
	break;
    case OUTPUT_MESSAGES:
    case OUTPUT_FILES:
	ret = do_search_messages (format, query, output, offset, limit, dupe);
	break;
    case OUTPUT_TAGS:
	ret = do_search_tags (notmuch, format, query);
	break;
    }

    notmuch_query_destroy (query);
    notmuch_database_destroy (notmuch);

    talloc_free (format);

    return ret;
}
